### ISSUE TITLE
SourceLocation line number off by one

### ISSUE THREAD

**Issue by rsmckinney** [graphql-java/graphql-java#1512]
**Describe the bug**
The line number stored in `SourceLocation` is off by one when obtained from `GraphqlAntrlToLanguage#getSourceLocation()`:
```java
    protected SourceLocation getSourceLocation(Token token) {
        MultiSourceReader.SourceAndLine sourceAndLine = multiSourceReader.getSourceAndLineFromOverallLine(token.getLine());
        int column = token.getCharPositionInLine() + 1;
        // graphql spec says line numbers start at 1
        int line = sourceAndLine.getLine() + 1;
        return new SourceLocation(line, column, sourceAndLine.getSourceName());
    }
```
An Antlr token has a line number starting with 1, therefore there is no need to add one in this method.

**To Reproduce**
Use the API to get the `SourceLocation` of any `Node`, the line consistently greater than it should be.

>On further inspection there are several other places where line numbers are off by one.  This commit e310dc3f appears to be the culprit.  Maybe consider auditing that change and backout the parts that add one to the token line, which is already one-based.

**Comment by bbakerman** [comment]
Thanks for reporting this.  We had other reports from people that its off by one.

I guess the Antlr doco is definitive

https://www.antlr.org/api/Java/org/antlr/v4/runtime/TokenSource.html#getLine()

> Get the line number for the current position in the input stream. The first line in the input is line 1.

We will fix this

**Comment by bbakerman** [comment]
I think part of the problem is that the MultiSourceReader that I built (take N sources and pretend its one) is zero based and graphql and antlr is ones based and I think its got confused

```
SourceAndLine sourceAndLine = multiSourceReader.getSourceAndLineFromOverallLine(token.getLine());
        int column = token.getCharPositionInLine();

        // graphql spec says line numbers start at 1
        SourceLocation sourceLocation = new SourceLocation(sourceAndLine.getLine()+1, column, sourceAndLine.getSourceName());
```

should be

```
SourceAndLine sourceAndLine = multiSourceReader.getSourceAndLineFromOverallLine(token.getLine()-1);
        int column = token.getCharPositionInLine();

        // graphql spec says line numbers start at 1
        SourceLocation sourceLocation = new SourceLocation(sourceAndLine.getLine()+1, column, sourceAndLine.getSourceName());

```

I need to write more tests to get this right

**Comment by rsmckinney** [comment]
Looks good.  It may be worthwhile to make your own `LineNumberReader` that is 1-based and avoid the 0 vs. 1 confusion in the codebase.  Just a thought.

I'd rather not release a version of my project with a snapshot build of graphql-java; do you plan to have a revised release soon?  Thanks.

**Comment by grrttedwards** [comment]
@rsmckinney The PR linked has a milestone tag of `13.0`, so looks like we will be waiting

**Comment by rsmckinney** [comment]
@grrttedwards thanks for the heads up. Instead of using a snapshot release my code checks for the bug and adjusts accordingly.

**Comment by andimarek** [comment]
@rsmckinney  @grrttedwards it will be released in 13.0 which will not happen very soon. (couple of weeks at least)

**Comment by realityforge** [comment]
I believe this issue can be closed as the associated PR has been merged
